### PR TITLE
[Draft] Added Validators to CustomActionParameter

### PR DESCRIPTION
**Pull request by Diamekod0221**
## Describe your changes
Add parameter validators to CustomAction. Expose an endpoint for CustomAction validation. Check if we need new Validators for CustomActions. Hook up validators to CustomActions as currently needed.

## Checklist before merge
- [ ] Related issue ID is placed at the beginning of PR title in \[brackets\] (can be GH issue or Nu Jira issue)
- [ ] Code is cleaned from temporary changes and commented out lines
- [ ] Parts of the code that are not easy to understand are documented in the code
- [ ] Changes are covered by automated tests
- [ ] Showcase in dev-application.conf added to demonstrate the feature
- [ ] Documentation added or updated
- [ ] Added entry in _Changelog.md_ describing the change from the perspective of a public distribution user
- [ ] Added _MigrationGuide.md_ entry in the appropriate subcategory if introducing a breaking change
- [ ] Verify that PR will be squashed during merge

### CODE REVIEW

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/ManagementApiEndpoints.scala [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import sttp.tapir.json.circe.jsonBody
+import sttp.tapir.derevo.schema
+
+class ManagementApiEndpoints(auth: EndpointInput[AuthCredentials]) extends BaseEndpointDefinitions {
+
+  private lazy val baseProcessManagementEndpoint = baseNuApiEndpoint.in("processManagement")
+
+  lazy val customActionValidationEndpoint: SecuredEndpoint[
+    (ProcessName, CustomActionRequest),
+    CustomActionValidationError,
```
Comment (gskrobisz) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/api/ManagementApiEndpoints.scala:31]: See comments in https://github.com/TouK/nussknacker/commit/7cd849ac0b77e2b0cdae606a84c2757b4fca66e9

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/ManagementResources.scala [hunk 2]
```diff
@@ -27,6 +27,7 @@ import pl.touk.nussknacker.ui.process.deployment.{DeploymentManagerDispatcher, D
 import pl.touk.nussknacker.ui.process.processingtype.ProcessingTypeDataProvider
 import pl.touk.nussknacker.ui.process.test.{RawScenarioTestData, ResultsWithCounts, ScenarioTestService}
 import pl.touk.nussknacker.ui.security.api.LoggedUser
+import pl.touk.nussknacker.ui.validation.CustomActionValidator
```
Comment (gskrobisz) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/api/ManagementResources.scala:30]: Seems to be not used.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala [hunk 3]
```diff
@@ -482,6 +482,7 @@ object NodesApiEndpoints {
       final case object NoPermission                                    extends NodesError with CustomAuthorizationError
       final case class MalformedTypingResult(msg: String)               extends NodesError
 
+      // FIXME: what does this message mean??
```
Comment (gskrobisz) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/api/NodesApiEndpoints.scala:485]: Please, remove comment. I guess here we need one-way serialization only to communicate exceptions.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala [hunk 4]
```diff
@@ -148,7 +157,7 @@ class DeploymentServiceImpl(
   private def validateDeploymentComment(comment: Option[String]): Future[Option[DeploymentComment]] =
     DeploymentComment.createDeploymentComment(comment, deploymentCommentSettings) match {
       case Valid(deploymentComment) => Future.successful(deploymentComment)
-      case Invalid(exc)             => Future.failed(ValidationError(exc.getMessage))
+      case Invalid(exc)             => Future.failed(CustomActionValidationError(exc.message))
```
Comment (gskrobisz) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala:160]: Use CustomActionValidator.validate here to validate deployment comment.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala [hunk 5]
```diff
@@ -593,13 +603,25 @@ class DeploymentServiceImpl(
           params
         )
         customActionOpt = manager.customActionsDefinitions.find(_.name == actionName)
-        _ <- existsOrFail(customActionOpt, CustomActionNonExisting(actionName))
+        _ <- existsOrFail(customActionOpt, CustomActionNonExistingError(s"${actionName.value} doesn't exist"))
+        validator = new CustomActionValidator(manager.customActionsDefinitions)
+        // TODO: remove this validation prosthesis
```
Comment (gskrobisz) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala:608]: Please remove this validation prosthesis and use action parameters validation.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/server/AkkaHttpBasedRouteProvider.scala [hunk 6]
```diff
@@ -391,11 +399,12 @@ class AkkaHttpBasedRouteProvider(
...
           appApiHttpService,
           componentsApiHttpService,
-          userApiHttpService,
+          managementApiHttpService,
+          nodesApiHttpService,
           notificationApiHttpService,
           scenarioActivityApiHttpService,
           scenarioParametersHttpService,
-          nodesApiHttpService
+          userApiHttpService
```
Comment (gskrobisz) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/server/AkkaHttpBasedRouteProvider.scala:407]: Why the order is changed?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+import pl.touk.nussknacker.engine.deployment.{
+  CustomActionDefinition,
+  CustomActionParameter,
+  CustomActionValidationResult
+}
+import pl.touk.nussknacker.engine.graph.expression.Expression
+import pl.touk.nussknacker.restmodel.CustomActionRequest
+import pl.touk.nussknacker.ui.BadRequestError
+
+class CustomActionValidator(val allowedActions: List[CustomActionDefinition]) {
```
Comment (gskrobisz) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:18]: Lets rename to ScenarioActionParameterValidator, we are aiming here to validation at "action parameter" level. And because of that I recommend to pass single action properties like this:
```
class ScenarioActionParameterValidator(actionParameters: Map[String, CustomActionParameter])
```

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+      checkedParams
+    )
+  }
+
+}
+
+object CustomActionValidationError {
+  def apply(message: String): CustomActionValidationError = new CustomActionValidationError(message)
+
+  implicit val encodeCustomActionValidationError: Encoder[CustomActionValidationError] =
```
Comment (gskrobisz) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:144]: This should be in ManagementApiEndpoint

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+}
+import pl.touk.nussknacker.engine.graph.expression.Expression
+import pl.touk.nussknacker.restmodel.CustomActionRequest
+import pl.touk.nussknacker.ui.BadRequestError
+
+class CustomActionValidator(val allowedActions: List[CustomActionDefinition]) {
+
+  def validateCustomActionParams(
+      request: CustomActionRequest
+  ): Either[CustomActionValidationError, CustomActionValidationResult] = {
```
Comment (gskrobisz) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:22]: Too much happens here. It seems that with `Either[CustomActionValidationError, CustomActionValidationResult]` you have 2 levels of error handling:
- Left(CustomActionValidationError) is returned when validation errors are detected
- Right(CustomActionValidationResult) contains validation errors.

Here we should return something simple like:
```
CustomActionValidationResult(parameterErrors: List[NodeValidationError])
```
In the code below we should handle validation error as Validated/ValidatedNel, instead of exception (see CustomActionValidationError and MismatchedParamsError). Also, keep in mind that:
- no validation error = ok, goes straight to 200 OK
- some validation errors = still ok, we should return list of invalid fields with reason of invalidity
- technical exception = not ok, goes straight to 500 server error (?)

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+                .collect { case Invalid(i) => i }
+            case None =>
+              throw new IllegalStateException
+          }
+        )
+      }
+    }
+  }
+
+  private def checkForMissingKeys(
```
Comment (gskrobisz) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:92]: See similar validation in ScenarioPropertiesValidator

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+        val customActionParams      = foundAction.parameters
+        val requestParamsMap        = getRequestParamsMap(request, customActionParams)
+
+        val validated = validateParams(requestParamsMap, customActionParams)
+        getValidationResult(validated)
+      }
+    }
+  }
+
+  private def getRequestParamsMap(request: CustomActionRequest, customActionParams: List[CustomActionParameter]) = {
```
Comment (gskrobisz) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:43]: See similar validation in `ScenarioPropertiesValidator`, it seems that here you want do `getMissingRequiredPropertyValidationResults` and `getUnknownPropertyValidationResults`.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+    hasErrors match {
+      case Right(true) =>
+        Right(CustomActionValidationResult.Invalid(validatedParams.getOrElse(throw new IllegalStateException)))
+      case Right(false) => Right(CustomActionValidationResult.Valid)
+      case Left(va)     => Left(va)
+    }
+  }
+
+  def validateCustomActionParams(
```
Comment (gskrobisz) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:119]: `validateCustomActionParams` is not in the responsibliity scope of validator, should be moved to DeploymentServiceImpl.

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ManagementResourcesSpec.scala [hunk 13]
```diff
@@ -421,8 +421,8 @@ class ManagementResourcesSpec
       ProcessTestData.sampleProcessName,
       CustomActionRequest(ScenarioActionName("non-existing"))
     ) ~> check {
-      status shouldBe StatusCodes.NotFound
-      responseAs[String] shouldBe "non-existing is not existing"
+      status shouldBe StatusCodes.BadRequest
```
Comment (gskrobisz) [inline designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ManagementResourcesSpec.scala:424]: Non existing action should be "not found".

File: extensions-api/src/main/scala/pl/touk/nussknacker/engine/deployment/CustomActionDefinition.scala [hunk 14]
```diff
@@ -26,7 +28,13 @@ case class CustomActionDefinition(
...
 
-//TODO: validators, defaultValue, hint, labelOpt?
-case class CustomActionParameter(name: String, editor: ParameterEditor)
+case class CustomActionParameter(name: String, editor: ParameterEditor, validators: Option[List[ParameterValidator]])
 
 case class CustomActionResult(msg: String)
+
+sealed trait CustomActionValidationResult
+
+object CustomActionValidationResult {
```
Comment (gskrobisz) [inline extensions-api/src/main/scala/pl/touk/nussknacker/engine/deployment/CustomActionDefinition.scala:37]: You should use `Validated` and `ValidatedNel` where Valid and Invalid are already implemented.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/ManagementResources.scala (hunk 2 above)
Comment (Diamekod0221) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/api/ManagementResources.scala]: imports updated

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala (hunk 5 above)
Comment (Diamekod0221) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala]: Used Validator

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/server/AkkaHttpBasedRouteProvider.scala (hunk 6 above)
Comment (Diamekod0221) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/server/AkkaHttpBasedRouteProvider.scala:407]: I have alligned components alphabetically to easily check order.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala (hunk 7 above)
Comment (Diamekod0221) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala]: Changed to injecting single aciton into the constructor. I think the naming is ok for now, we can clean up when the change to action parameter comes

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala (hunk 9 above)
Comment (Diamekod0221) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala]: Extracted error logic to ManagementApiHttpService, CustomActionValidator now returns ValidatedNel[SubGraphCompilationError, Unit]

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala (hunk 10 above)
Comment (Diamekod0221) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala]: removed

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala (hunk 11 above)
Comment (Diamekod0221) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala]: added a checkForMissingParameters

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala (hunk 12 above)
Comment (Diamekod0221) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:115]: moved

Comment (gskrobisz) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:115]: I still see here `validateCustomActionParams`and `fromCommand`. Lets keep that validator responsible only for validation on action level, and move that translation from command outside, to `DeploymentServiceImpl`.

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ManagementResourcesSpec.scala [hunk 15]
```diff
@@ -422,7 +422,9 @@ class ManagementResourcesSpec
       CustomActionRequest(ScenarioActionName("non-existing"))
     ) ~> check {
       status shouldBe StatusCodes.NotFound
-      responseAs[String] shouldBe "non-existing is not existing"
+      responseAs[
+        String
+      ] shouldBe s"Couldn't find definition of action non-existing for scenario ${ProcessTestData.sampleProcessName} when trying to validate"
```
Comment (gskrobisz) [inline designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ManagementResourcesSpec.scala:427]: Maybe "Couldn't find definition of action non-existing" is sufficient?

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/validation/CustomActionValidatorTest.scala [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+      "testStatus1" :: "testStatus2" :: Nil,
+      testCustomActionParams
+    )
+
+  private val noParamsCustomAction =
+    CustomActionDefinition(ScenarioActionName("noparams"), "testStatus" :: Nil, Nil)
+
+  private val validator = new CustomActionValidator(testCustomAction)
+
+  type TestParams = Map[String, String]
```
Comment (gskrobisz) [inline designer/server/src/test/scala/pl/touk/nussknacker/ui/validation/CustomActionValidatorTest.scala:48]: It it used only in one place. Seems to be obsolete.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+import pl.touk.nussknacker.engine.api.NodeId
+import pl.touk.nussknacker.engine.api.context.PartSubGraphCompilationError
+import pl.touk.nussknacker.engine.api.context.ProcessCompilationError.MismatchParameter
+import pl.touk.nussknacker.engine.api.definition.MandatoryParameterValidator
+import pl.touk.nussknacker.engine.api.deployment.CustomActionCommand
+import pl.touk.nussknacker.engine.api.parameter.ParameterName
+import pl.touk.nussknacker.engine.deployment.{CustomActionDefinition, CustomActionParameter}
+import pl.touk.nussknacker.engine.graph.expression.Expression
+import pl.touk.nussknacker.restmodel.CustomActionRequest
+import pl.touk.nussknacker.ui.{BadRequestError, NotFoundError}
```
Comment (gskrobisz) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:14]: Unused import {BadRequestError, NotFoundError}

File: designer/client/src/components/modals/CustomActionDialog.tsx [hunk 18]
```diff
@@ -38,6 +40,28 @@ function CustomActionForm(props: CustomActionFormProps): JSX.Element {
...
+    const validationRequest: CustomActionValidationRequest = {
+        actionName: action.name,
+        params: state,
+    };
+    const debouncedValidateCustomAction = debounce(async () => {
+        const data = await HttpService.validateCustomAction(processName, validationRequest);
+        setErrors(data.validationErrors);
+    }, 300);
+
+    useEffect(() => {
```
Comment (Dzuming) [inline designer/client/src/components/modals/CustomActionDialog.tsx:55]: You can merge, these useEffect, and update dependencies

```
 useEffect(() => {
        debouncedValidateCustomAction();

      return () => {
        debouncedValidateCustomAction.cancel();
      };
    }, [debouncedValidateCustomAction]);
```

File: designer/client/src/components/modals/CustomActionDialog.tsx [hunk 19]
```diff
@@ -38,6 +40,28 @@ function CustomActionForm(props: CustomActionFormProps): JSX.Element {
...
     useEffect(() => onChange(state), [onChange, state]);
 
+    const [errors, setErrors] = useState<NodeValidationError[]>([]);
+
+    const processName = useSelector(getProcessName);
+    const validationRequest: CustomActionValidationRequest = {
+        actionName: action.name,
+        params: state,
+    };
+    const debouncedValidateCustomAction = debounce(async () => {
```
Comment (Dzuming) [inline designer/client/src/components/modals/CustomActionDialog.tsx:50]: You can memoize this action

```
const debouncedValidateCustomAction = useMemo(
        () =>
            debounce(async () => {
                const data = await HttpService.validateCustomAction(processName, validationRequest);
                setErrors(data.validationErrors);
            }, 300),
        [processName, validationRequest],
    );
```

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/validation/CustomActionValidatorTest.scala [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+    CustomActionDefinition(
+      testCustomActionName,
+      "testStatus1" :: "testStatus2" :: Nil,
+      testCustomActionParams
+    )
+
+  private val noParamsCustomAction =
+    CustomActionDefinition(ScenarioActionName("noparams"), "testStatus" :: Nil, Nil)
+
+  private val validator = new CustomActionValidator(testCustomAction)
```
[the remark this answers is not in the document]
Comment (Diamekod0221) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/validation/CustomActionValidatorTest.scala:46]: The ones that create their own validator use some custom validator state. I could maybe rename to customValidator for clarity and define the whole setup block outside of test scope for consistency?

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ManagementResourcesSpec.scala (hunk 15 above)
Comment (Diamekod0221) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ManagementResourcesSpec.scala:427]: I think it is good to write the scenario name here to make sure that we are actually processing what we want. More information shouldn't hurt :)

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/validation/CustomActionValidatorTest.scala (hunk 20 above)
[the remark this answers is not in the document]
Comment (Diamekod0221) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/validation/CustomActionValidatorTest.scala:46]: I have renamed to `defaultValidator` and `customValidator` for clarity. There's only one test that uses a custom validator so i think it's okay to use the current structure.

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/validation/CustomActionValidatorTest.scala (hunk 16 above)
Comment (Diamekod0221) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/validation/CustomActionValidatorTest.scala:48]: removed

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ManagementResourcesSpec.scala (hunk 15 above)
Comment (gskrobisz) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ManagementResourcesSpec.scala:427]: Ok, lets keep the scenario name. The part "when trying to validate" is not true. With current implementation it's more like "when trying to execute an action" and is obsolete.

File: designer/client/src/components/modals/CustomActionDialog.tsx (hunk 19 above)
Comment (Dzuming) [reply designer/client/src/components/modals/CustomActionDialog.tsx]: We discussed it internally, and setTimeout will fit our needs in this case

**Review by Dzuming — APPROVED** [review]

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/ManagementApiHttpService.scala [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+            scenarioIdWithName = ProcessIdWithName(scenarioId, processName)
+            actionDefinition <- getActionDefinition(scenarioIdWithName, req.actionName)
+            validator = new CustomActionValidator(actionDefinition)
+            resultsDto <- getValidationResultsDto(validator, req)
+          } yield resultsDto
+        }
+      }
+  }
+
+  private def getScenarioIdByName(scenarioName: ProcessName): EitherT[Future, NoScenario, ProcessId] = {
```
Comment (arkadius) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/api/ManagementApiHttpService.scala:48]: I extracted a helper trait for that: `ScenarioHttpServiceExtensions` to avoid code duplication. Can you reuse it?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+import pl.touk.nussknacker.engine.api.context.PartSubGraphCompilationError
+import pl.touk.nussknacker.engine.api.context.ProcessCompilationError.MismatchParameter
+import pl.touk.nussknacker.engine.api.definition.MandatoryParameterValidator
+import pl.touk.nussknacker.engine.api.deployment.CustomActionCommand
+import pl.touk.nussknacker.engine.api.parameter.ParameterName
+import pl.touk.nussknacker.engine.deployment.{CustomActionDefinition, CustomActionParameter}
+import pl.touk.nussknacker.engine.graph.expression.Expression
+import pl.touk.nussknacker.restmodel.CustomActionRequest
+
+class CustomActionValidator(allowedAction: CustomActionDefinition) {
```
Comment (arkadius) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:15]: Do you see an option to unify this with `ScenarioPropertiesValidator`? It is a huge duplication of quite easy to broken business logic. Also users may be disgusted if they find out that this logic behave different for some corner cases.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+            s"Couldn't find a matching parameter in action definition for this param: $name",
+            "",
+            ParameterName(name),
+            nodeId.id
+          ).invalidNel[Unit]
+      }
+    }.sequence_
+  }
+
+  private def toValidatedNel(param: CustomActionParameter, expressionValue: String, parameterName: String)(
```
Comment (arkadius) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:78]: This name is confusing. It looks like you do some dumb conversion here but in fact, here is done the core of the logic - validations

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+      customActionParams: List[CustomActionParameter]
+  )(implicit nodeId: NodeId): ValidatedNel[PartSubGraphCompilationError, Unit] = {
+
+    customActionParams
+      .collect { case param if param.validators.nonEmpty => (param.name, param.validators) }
+      .flatMap { case (name, validators) =>
+        validators.collect { case validator => validator }.map {
+          _.isValid(
+            paramName = ParameterName(name),
+            expression = Expression.spel("None"),
```
Comment (arkadius) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:104]: What the heck? Why do you put some `None` into spel expression?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala [hunk 25]
```diff
@@ -0,0 +1,129 @@
...
+
+  private def validateParams(
+      requestParamsMap: Map[String, String],
+      customActionParams: List[CustomActionParameter]
+  )(implicit nodeId: NodeId): ValidatedNel[PartSubGraphCompilationError, Unit] = {
+
+    requestParamsMap match {
+      case paramsMap if paramsMap.nonEmpty =>
+        handleNonEmptyParamsRequest(paramsMap, customActionParams)
+      case emptyMap if emptyMap.isEmpty =>
```
Comment (arkadius) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:54]: Why the empty map is some special case? Eventually you traverse all params so what's the difference?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala [hunk 26]
```diff
@@ -646,17 +649,33 @@ class DeploymentServiceImpl(
...
         )
         customActionOpt = manager.customActionsDefinitions.find(_.actionName == actionName)
-        _ <- existsOrFail(customActionOpt, CustomActionNonExisting(actionName))
+        customAction <- existsOrFail(
+          customActionOpt,
+          CustomActionNonExistingError(
+            s"Couldn't find definition of action ${actionName.value} for scenario ${processIdWithName.name} when trying to validate"
+          )
+        )
+        _ <- validateActionCommand(actionCommand, customAction)
```
Comment (arkadius) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala:658]: I can't see any test covering this important case

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala [hunk 27]
```diff
@@ -646,17 +649,33 @@ class DeploymentServiceImpl(
...
         _ = checkIfCanPerformActionInState(actionName, processDetails, processState)
         invokeActionResult <- DBIOAction.from(manager.processCommand(actionCommand))
       } yield invokeActionResult
     )
   }
 
+  private def validateActionCommand(actionCommand: CustomActionCommand, customAction: CustomActionDefinition) = {
+    val validator        = new CustomActionValidator(customAction)
+    val validationResult = validator.validateCustomActionParams(actionCommand)
+    val validationFlag   = validationResult
```
Comment (arkadius) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala:668]: Flag? Flag is a boolean parameter. It:
- isn't boolean
- isn't parameter - it's the result

Also, this local variable doesn't add any value, you already have `validationResult`

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala (hunk 22 above)
Comment (Diamekod0221) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:15]: I think in the current state we could possibly merge these two in a separate refactor. I don't think that's a good idea, because ScenarioProperties is a different object from CustomAction. I think merging this makes the code unnecesarily complex, especially if it turns out that the logic isn't exactly the same or we have to change it in the future. It's just 90 extra lines. 
Your user argument is convincing, but we should just make sure the external results are the same.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala (hunk 25 above)
Comment (Diamekod0221) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:54]: In case that the map is empty I'm checking if there are validators that will pop off for an empty request (like MandatoryParameter). Because of this i have divided this check into two cases.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala (hunk 26 above)
Comment (Diamekod0221) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala:658]: Please specify which line do you mean. This: 
` customAction <- existsOrFail(
          customActionOpt,
          CustomActionNonExistingError(
            s"Couldn't find definition of action ${actionName.value} for scenario ${processIdWithName.name} when trying to validate"
          )
        )
        ` ?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala (hunk 27 above)
Comment (Diamekod0221) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala:668]: Tru, leftover code. Fixed.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala (hunk 24 above)
Comment (Diamekod0221) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:104]: It's an attempt to validate an empty expression when the user doesn't input anything. Is there some more subtle way?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala (hunk 23 above)
Comment (Diamekod0221) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala]: Changed the name.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/services/ManagementApiHttpService.scala [hunk 28]
```diff
@@ -0,0 +1,83 @@
...
+            scenarioIdWithName = ProcessIdWithName(scenarioId, processName)
+            actionDefinition <- getActionDefinition(scenarioIdWithName, req.actionName)
+            validator = new CustomActionValidator(actionDefinition)
+            resultsDto <- getValidationResultsDto(validator, req)
+          } yield resultsDto
+        }
+      }
+  }
+
+  private def getScenarioIdByName(
```
[the remark this answers is not in the document]
Comment (Diamekod0221) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/services/ManagementApiHttpService.scala]: I don't want to use it because it would effectively replace 3 lines of code with 7 or more and add complexity.

[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/services/ManagementApiHttpService.scala]: What plan do you recommend to deal with this code duplication (getting scenario details, checking if scenario exists)?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala (hunk 24 above)
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:104]: I don't understand your answer. Why do you validate an empty expression when the user doesn't input anything?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala (hunk 25 above)
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:54]: And for one element map with missing some parameter value that is mandatory you don't need to do the same?

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala (hunk 26 above)
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala:658]: No, this:
```
DBIOAction.failed(new IllegalStateException(s"Validation failed for: ${actionCommand.actionName}"))
    }
```

Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala:658]: Also IMO it shouldn't be 500, it is bad request - not an illegal state of the application

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/services/ManagementApiHttpService.scala (hunk 28 above)
[the remark this answers is not in the document]
Comment (Diamekod0221) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/services/ManagementApiHttpService.scala]: I think your trait is a good idea, but in this case it's a one liner 
`
            scenarioId <- EitherT.fromOptionF(processService.getProcessId(processName), NoScenario(processName))`
            that doesn't make use of all the functionality the trait provides and is basically a wrapped function call. Adding some extra code and complexity on top of it doesn't make that much sense to me

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala (hunk 25 above)
Comment (Diamekod0221) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala:54]: you are right, I forgot to clean this up when i extracted 
`private def checkForMissingParams(
      requestParamsMap: Map[String, String],
      customActionParams: List[CustomActionParameter]`
      This emptyParamsMapRequest can probably go along with the None expression check. I will update it.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala (hunk 26 above)
Comment (Diamekod0221) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala:658]: I have changed the error type and added a test case.

**Review by arkadius — APPROVED** [review]
one comment left. fix it and you can merge
